C26 updates to Unit 4 PSE 5 (#21)
Ada-Developers-Academy/cs-fun-interview-practice: 05/network-delay.md

<br>

One thing the feedback called out as a strength was keeping the "convert the edge list into an adjacency dict" step separate from the Dijkstra step itself, rather than folding the conversion logic into the traversal. The interviewer said this made it clear I recognized Dijkstra needs a different data shape than the raw input, without getting into implementation details about dictionaries or lists. I think this landed well because it showed I was thinking about data representation as its own concern, not just as a side effect of "doing the algorithm."

<br>

The main improvement suggested was to say more about why a priority queue is required rather than just listing it as one of the data structures I'm initializing. Right now my steps mention it alongside the visited set and cost list without explaining its role. If I revised this, I'd add a sentence explaining that the priority queue is what lets us always process the closest unvisited node next, which is the property that guarantees a node's cost is finalized the first time we visit it — otherwise it's not obvious why a plain queue or list wouldn't work just as well.

<br>

> One suggestion I pushed back on was a recommendation to add a step for handling negative edge weights before running Dijkstra. That doesn't apply here — the problem defines wᵢ as a signal travel time, which can't be negative, and the hints for this problem already establish that assumption as the reason Dijkstra's algorithm is a valid choice in the first place. Adding a check for something the problem statement rules out by definition would be solving for a case that can't occur, so I left that out.

Reviewer: The last sentence feels a little weird, I'd consider either removing the sentence or adding a few words for clarity.

```suggestion
One suggestion I pushed back on was a recommendation to add a step for handling negative edge weights before running Dijkstra. That doesn't apply here — the problem defines wᵢ as a signal travel time, which can't be negative, and the hints for this problem already establish that assumption as the reason Dijkstra's algorithm is a valid choice in the first place. Adding a check for something the problem statement rules out would be solving for a case that can't occur, so I left that out of the implementation.
```

<br>

> The part that surprised me was a comment pointing out that my final step, "return the maximum cost from the minimum cost list," is really doing two different jobs at once: confirming reachability (are we sure every node got a cost?) and then computing the answer (what's the slowest of those costs?). I had written it as a single step, but separating those two ideas out loud helped me see more clearly why the problem asks for a "maximum of minimums" instead of a sum or an average — it's because the overall delay is determined by whichever node is slowest to hear the signal, not by the total work done to reach everyone.

Reviewer: Not necessary to change, but the "out loud" in the middle sounded a little weird, so I looked at how we could shorten the text a little while I was in there. 

```suggestion
I was surprised by a comment pointing out that my final step, "return the maximum cost from the minimum cost list," is doing two jobs at once: confirming reachability (are we sure every node got a cost?) and then computing the answer (what's the slowest of those costs?). I wrote it as a single step, but separating those two ideas out clarified for me why the problem asks for a "maximum of minimums" instead of a sum or an average. It's because the overall delay is determined by whichever node is slowest to hear the signal, not by the total work done to reach everyone.
```
